Octokit/Clients/Enterprise/EnterpriseProbe.cs

```diff
@@ -0,0 +1,97 @@
+﻿using System;
+using System.Net;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Octokit.Internal;
+
+namespace Octokit
+{
+    /// <summary>
+    /// Service used to check if a GitHub Enterprise instance lives at a given URL.
+    /// </summary>
+    public class EnterpriseProbe : IEnterpriseProbe
+    {
+        static readonly Uri endPoint = new Uri("/site/sha", UriKind.Relative);
+        readonly ProductHeaderValue productHeader;
+        readonly IHttpClient httpClient;
+
+        public EnterpriseProbe(ProductHeaderValue productHeader, IHttpClient httpClient)
+        {
+            Ensure.ArgumentNotNull(productHeader, "productHeader");
+            Ensure.ArgumentNotNull(httpClient, "httpClient");
+
+            this.productHeader = productHeader;
+            this.httpClient = httpClient;
+        }
+
+        /// <summary>
+        /// Makes a request to the specified URL and returns whether or not the probe could definitively determine that a GitHub
+        /// Enterprise Instance exists at the specified URL.
+        /// </summary>
+        /// <remarks>
+        /// The probe checks the absolute path /site/sha at the specified <paramref name="enterpriseBaseUrl" />.
+        /// </remarks>
+        /// <param name="enterpriseBaseUrl">The URL to test</param>
+        /// <returns>An <see cref="EnterpriseProbeResult" /> with either <see cref="EnterpriseProbeResult.Ok"/>,
+        /// <see cref="EnterpriseProbeResult.NotFound"/>, or <see cref="EnterpriseProbeResult.Failed"/> in the case the request failed</returns>
+        public async Task<EnterpriseProbeResult> Probe(Uri enterpriseBaseUrl)
+        {
+            var request = new Request
+            {
+                Method = HttpMethod.Get,
+                BaseAddress = enterpriseBaseUrl,
+                Endpoint = endPoint,
+                Timeout = TimeSpan.FromSeconds(3)
+            };
+            request.Headers.Add("User-Agent", productHeader.ToString());
+
+            try
+            {
+                var response = await httpClient.Send(request);
+                return response == null
+                    ? EnterpriseProbeResult.Failed
+                    : (IsEnterpriseResponse(response)
+                        ? EnterpriseProbeResult.Ok
+                        : EnterpriseProbeResult.NotFound);
+            }
+            catch (Exception)
+            {
+                return EnterpriseProbeResult.Failed;
```

Comment:
What I found while poking at enterprise responses is that if it throws an exception, it might still be an enterprise server, it just might be configured to not let you poke without authentication, even if this entry point is supposed to be unauthenticated. In this case, the exception will have the response, so this can look for that (I do it like so [here](https://github.com/github/VisualStudio/blob/74ee1705e088f4250b5ba195046f48ce0ab2e9fa/src/GitHub.Exports/ExceptionExtensions.cs))